src/apps/invitations/services.py

```diff
@@ -56,6 +58,45 @@ def __init__(self, session, user: User):
     async def fetch_all(self, query_params: QueryParams) -> list[InvitationDetail]:
         return await self.invitations_crud.get_pending_by_invitor_id(self._user.id, query_params)
 
+    async def fetch_by_emails(self, emails: list[str]) -> list[InvitationDetail]:
+        return await self.invitations_crud.get_latest_by_emails(emails)
+
+    async def fill_pending_invitations_respondents(
+        self, respondents: list[WorkspaceRespondent]
+    ) -> list[WorkspaceRespondent]:
+        emails = [respondent.email for respondent in respondents if respondent.email is not None]
+        invitations = await self.fetch_by_emails(emails)
+
```

Comment:
You can return here a map with email_applet_id-> invitation, so bellow just get invitation using invitations[email_applet_id], this way you don't need 3 for loops

Reply:
Thanks for the suggestion 👌